app/lib/state_diagram.rb

```diff
@@ -0,0 +1,51 @@
+class StateDiagram
+  def self.svg(only_from_state = nil)
+    graph = GraphViz.new('G', rankdir: 'TB', ratio: 'fill')
+
+    states_to_show = []
+
+    ApplicationStateChange.workflow_spec.states.each do |_, state|
+      next if only_from_state && state.name != only_from_state.to_sym
+
+      state.events.flat.each do |event|
+        states_to_show << state.name
+        states_to_show << event.transitions_to
+
+        graph.add_edges(state.name.to_s, event.transitions_to.to_s, label: event_name(state.name, event), fontname: 'Helvetica')
+      end
+    end
+
+    states_to_show.flatten!
+    states_to_show.uniq!
+
+    ApplicationStateChange.workflow_spec.states.each do |state_name, state|
+      if only_from_state
+        next unless only_from_state.to_sym == state.name || state.name.to_sym.in?(states_to_show)
+      end
+
+      graph.add_nodes(
+        state_name.to_s,
+        label: I18n.t!("application_states.#{state_name}.name"),
+        width: '1',
+        height: '1',
+        shape: 'rect',
+        fontname: 'Helvetica',
+      )
+    end
+
+    graph.output(svg: String)
+  end
+
+  def self.event_name(state, event)
+    by = I18n.t!("events.#{state}-#{event}.by")
+
+    emoji = {
+      'candidate' => '👩‍🎓',
+      'referee' => '👩‍🏫',
+      'provider' => '🏫',
+      'system' => '🤖',
+    }.fetch(by)
```

Comment:
I nominate this for the best Ruby code I have read and probably will read this year